compiler/src/dmd/funcsem.d

```diff
@@ -2986,3 +2986,55 @@ extern (D) void checkMain(FuncDeclaration fd)
     if (retType.ty != Tint32 && retType.ty != Tvoid && retType.ty != Tnoreturn)
         .error(fd.loc, "%s `%s` must return `int`, `void` or `noreturn`, not `%s`", fd.kind, fd.toPrettyChars, tf.nextOf().toChars());
 }
+
+/***********************************************
+ * Check all return statements for a function to verify that returning
+ * using NRVO is possible.
+ *
+ * Returns:
+ *      `false` if the result cannot be returned by hidden reference.
+ */
+extern (D) bool checkNRVO(FuncDeclaration fd)
+{
+    if (!fd.isNRVO() || fd.returns is null)
+        return false;
+
+    auto tf = fd.type.toTypeFunction();
+    if (tf.isref)
+        return false;
+
+    foreach (rs; *fd.returns)
+    {
+        if (auto ve = rs.exp.isVarExp())
+        {
+            auto v = ve.var.isVarDeclaration();
+            if (!v || v.isReference())
+                return false;
+            else if (fd.nrvo_var is null)
+            {
+                // Variables in the data segment (e.g. globals, TLS or not),
+                // parameters and closure variables cannot be NRVOed.
+                if (v.isDataseg() || v.isParameter() || v.toParent2() != fd)
+                    return false;
+                if (v.nestedrefs.length && fd.needsClosure())
+                    return false;
+                // don't know if the return storage is aligned
+                version (MARS)
+                {
+                    if (fd.alignSectionVars && (*fd.alignSectionVars).contains(v))
+                        return false;
+                }
```

Comment:
`MARS` isn't defined anywhere in this module.

See the top of func.d https://github.com/dlang/dmd/blob/d1f5a7ebd251b3df263a3c6ff87cac4aa9be309c/compiler/src/dmd/func.d#L62-L64

This boilerplate will need to be copied over too.

Reply:
Cheers. Missed that locally---I'm compiling with LDC2 on a mac.